### PR TITLE
HeteroLinear/SEGMM: switch from heuristic to timing-cache

### PR DESCRIPTION
**Pull request by puririshi98**
needed to rebase https://github.com/pyg-team/pytorch_geometric/pull/8472

### CODE REVIEW

**Review by akihironitta — COMMENTED** [review]
What does this PR do on a high level?

File: torch_geometric/nn/dense/linear.py
Comment (akihironitta) [inline torch_geometric/nn/dense/linear.py:330]: ~noob question: What does "cses" stand for?~

```suggestion
            # TODO check cases of compiling and scripting properly
```

File: torch_geometric/nn/dense/linear.py
Comment (akihironitta) [inline torch_geometric/nn/dense/linear.py:282]: nit: I'm in favor of putting this as a constant outside the class so that one can easily experiment this.

File: CHANGELOG.md
Comment (akihironitta) [inline CHANGELOG.md:57]: I don't think timing-cache is clear to users. Maybe something like this?

```suggestion
- Changed `use_segment_matmul` in `HeteroLinear`, and it now benchmarks both with and without the flag and uses a faster one in subsequent calls ([#8615](https://github.com/pyg-team/pytorch_geometric/pull/8615))
```

**Review by akihironitta — COMMENTED** [review]
- Is this something similar to https://pytorch.org/docs/stable/backends.html#torch.backends.cudnn.benchmark in a sense that it first benchmarks all algos and then select the fastest one? Shall we make this as an opt-in feature just like `cudnn.benchmark`?
- I'm not sure how unlikely it is, but is the measurament robust enough for each rank to always pick up the same algorithm in a distributed setting?

File: torch_geometric/nn/dense/linear.py
Comment (puririshi98) [reply torch_geometric/nn/dense/linear.py:282]: agreed

**Comment by puririshi98** [comment]
> Is this something similar to https://pytorch.org/docs/stable/backends.html#torch.backends.cudnn.benchmark in a sense that it first benchmarks all algos and then select the fastest one? Shall we make this as an opt-in feature just like cudnn.benchmark?

i think in our case there are only two algorithms to check and in many cases the speed difference is drastic(in either direction). we are currently selecting it based on sklearn hueristics i trained from benchmarks on a single A100 machine on a single software stack from about half a year ago. this is far less robust on the majority of cards besides A100. i think the new solution is much better. I think instead of making the hueristic opt-in seperately, we can make it a part of the existing `torch_geometric.backend.use_segment_matmul

> I'm not sure how unlikely it is, but is the measurament robust enough for each rank to always pick up the same algorithm in a distributed setting?

i think this is a very low likelihood and we run a similarly low risk with the current heuristic when using dynamic shapes on each rank

**Comment by puririshi98** [comment]
@rusty1s any thoughts on the above discussion?

**Comment by rusty1s** [comment]
Yes, I agree that this solution is much better than the heuristic-based one. I think we would need to avoid measuring warmup times though, and likely adjust `MEASURE_ITER` based on whether we are in a test or in an actual example.

**Comment by puririshi98** [comment]
`
Yes, I agree that this solution is much better than the heuristic-based one. I think we would need to avoid measuring warmup times though, and likely adjust MEASURE_ITER based on whether we are in a test or in an actual example.

`

that sounds like a good idea to me. do you have a suggested method of implementing such a system?

**Comment by rusty1s** [comment]
Currently, we use `'pytest' in sys.modules` to detect whether we are testing.

**Comment by puririshi98** [comment]
okay thanks @rusty1s, do you have a desired value for MEASURE_ITER when we are testing? how about for when we aren't

**Comment by rusty1s** [comment]
I think for testing `MEASURE_ITER=1` would be sufficient. The best value for scripts needs to be recommended by you :)

**Review by akihironitta — APPROVED** [review]
LGTM! (but I'd suggest waiting for another review from Matthias)

Also, I think we can follow up in a separate PR, but we should update `RGCNConv` right?

**Comment by puririshi98** [comment]
`but we should update RGCNConv right?`
i agree, i can add this to my todo list and will follownup w/ a future PR

**Comment by puririshi98** [comment]
@rusty1s anything else needed to merge?

**Comment by rusty1s** [comment]
Please let me take a look tomorrow :) I'll try to get to it. If you are feeling blocked, please go ahead and merge.

**Review by rusty1s — APPROVED** [review]
That's nice :)

**Pull request merged**
